### PR TITLE
haskell.packages.ghc822/844: Fix configuration and build of GHC822/844

### PR DESCRIPTION
**Pull request by artempyanykh**
###### Motivation for this change

`nix-env -qaP -A haskell.packages.ghc8xx` for GHC822 and 844 fails with:

    nix called without required argument 'ghc-heap'

Adding ghc-heap = null to respective configuration files leads to further failures due to missing attributes (specific version of haskell packages).
Furthermore, when all configuration problems are resolved the ghc build itself fails due to sphinx.

I'm new to nix, so if there's a way to avoid copy-paste or version-pinning, please let me know!

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] Manjaro Linux 18.0.2
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/haskell-packages.nix [hunk 1]
```diff
@@ -55,7 +55,7 @@ in {
     };
     ghc844 = callPackage ../development/compilers/ghc/8.4.4.nix {
       bootPkgs = packages.ghc822Binary;
-      inherit (buildPackages.python3Packages) sphinx;
+      sphinx = buildPackages.python3Packages.sphinx_1_7_9;
```
Comment (dotlambda) [inline pkgs/top-level/haskell-packages.nix]:
```suggestion
      sphinx = buildPackages.python3Packages.sphinx.overridePythonAttrs (oldAttrs: rec {
        version = "1.7.9";
        src = oldAttrs.src.override {
          inherit version;
        };
      });
```

**Review by peti — CHANGES_REQUESTED** [review]
Please don't edit `pkgs/development/haskell-modules/hackage-packages.nix`. Please see the comment at the very top.

File: pkgs/top-level/haskell-packages.nix (hunk 1 above)
Comment (artempyanykh) [reply pkgs/top-level/haskell-packages.nix]: @dotlambda oh, that's cool! I'll remove sphinx/1.7.9.nix and apply your suggestion.

**Comment by artempyanykh** [comment]
@peti thanks, I see from @dotlambda 's suggestion how to change version/sha256 without creating new attributes in hackage-packages. I'll make the change.

**Comment by artempyanykh** [comment]
@peti looks like after all I haven't completely figured it out. A couple things:
1. Unfortunately,  I couldn't figure out how to call `overrideAttrs` within configuration-ghc-xx, so I just moved relevant definitions to respective configuration files out of hackage-packages. Even then, it's not particularly clear why we need some special version-pinned packages in a specific ghc configuration (for instance yaml_0_11_.. is no longer a thing in 844 config) or if overrideAttrs is any clearer than just dropping a full derivation.
2. Cabal_1_24_2_0 in 822 compiles with one more patch for dependency constraint.
3. hackage-db-2_0_1 fails during compilation --- for some reason aeson doesn't have a dependency on contravarian in hackage-packages which it should have.
4. stack and hpack_0_31_1 in 844 (after removing non-working yaml dep) started to compile... and still are compiling after several hours. Will update later.

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix [hunk 2]
```diff
@@ -43,9 +44,35 @@ self: super: {
   text = self.text_1_2_3_1;
 
   # Make sure we can still build Cabal 1.x.
-  Cabal_1_24_2_0 = overrideCabal super.Cabal_1_24_2_0 (drv: {
-    prePatch = "sed -i -e 's/process.*< 1.5,/process,/g' Cabal.cabal";
-  });
+  Cabal_1_24_2_0 = self.callPackage
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: What is the point of that change? I don't see what this is supposed to accomplish?

Comment (artempyanykh) [reply pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: @peti `super.Cabal_1_24_2_0` is not a thing anymore (I guess it was in hackage_packages sometime ago, but not now). Without this change `nix-env -qaP haskell.packages.ghc822` fails. 

I'd simply remove this derivation, but
1. the comment above says "# Make sure we can still build Cabal 1.x." 🤷‍♂️ 
2. removing this derivation is a breaking change _if_ someone depends on it and\or cares.

So, I fixed this `super.Cabal_1_24_2_0` issue and adjusted `prePatch` section so that Cabal successfully builds.

Does this answer your question?

P.S. If no one cares about `pkgname_{version}` derivations in ghc_8.x configurations, I can just remove them.

Comment (peti) [reply pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: Just drop the override. It's been broken for a long time and doubt anyone cares about Cabal 1.x any more.

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix [hunk 3]
```diff
@@ -89,7 +116,24 @@ self: super: {
   cabal2nix = super.cabal2nix.overrideScope (self: super: { Cabal = self.Cabal_2_2_0_1; });
   cabal2spec = super.cabal2spec.overrideScope (self: super: { Cabal = self.Cabal_2_2_0_1; });
   distribution-nixpkgs = super.distribution-nixpkgs.overrideScope (self: super: { Cabal = self.Cabal_2_2_0_1; });
-  hackage-db_2_0_1 = super.hackage-db_2_0_1.overrideScope (self: super: { Cabal = self.Cabal_2_2_0_1; });
+  hackage-db_2_0_1 = (self.callPackage
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: Why did you add this Nix build into the overrides file?

File: pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix [hunk 4]
```diff
@@ -64,12 +65,45 @@ self: super: {
   stack = (doJailbreak super.stack).override {
     Cabal = self.Cabal_2_4_1_0;
     hpack = self.hpack_0_31_1.override { Cabal = self.Cabal_2_4_1_0; };
-    yaml = self.yaml_0_11_0_0;
     hackage-security = self.hackage-security.override { Cabal = self.Cabal_2_4_1_0; };
   };
-  hpack_0_31_1 = super.hpack_0_31_1.override {
-    yaml = self.yaml_0_11_0_0;
-  };
+  hpack_0_31_1 = self.callPackage
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix]: Same here. Why did you add that build expression into the overrides file?

File: pkgs/top-level/haskell-packages.nix [hunk 5]
```diff
@@ -55,7 +55,13 @@ in {
     };
     ghc844 = callPackage ../development/compilers/ghc/8.4.4.nix {
       bootPkgs = packages.ghc822Binary;
-      inherit (buildPackages.python3Packages) sphinx;
+      sphinx = buildPackages.python3Packages.sphinx.overridePythonAttrs (oldAttrs: rec {
```
Comment (peti) [inline pkgs/top-level/haskell-packages.nix]: Please add a sphinx_1_7_9 attribute into `python-packages.nix` instead.

**Review by peti — CHANGES_REQUESTED** [review]

File: pkgs/top-level/haskell-packages.nix (hunk 5 above)
Comment (artempyanykh) [reply pkgs/top-level/haskell-packages.nix]: Done.

File: pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix (hunk 3 above)
Comment (artempyanykh) [reply pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix]: IDK what's the policy regarding keeping old version pinned derivations (and whether it's important to preserve them).

I assume that according to your comment about Cabal_1_x, no one cares about these, so I just removed them. Let me know if something other is desired.

File: pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix (hunk 4 above)
Comment (artempyanykh) [reply pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix]: Removed.

**Comment by artempyanykh** [comment]
* `nix-env -qaP -A haskell.packages.ghc8xx` works.
* ghc844 builds just fine.
* sphinx_1_7_9 moved under pyhon_packages.
* Old version pinned derivations in 822/844 configurations are removed.

@peti looks good?

**Review by peti — APPROVED** [review]

**Comment by peti** [comment]
Merged to `haskell-updates` as a251d91d2a0d1d5007bdbd13a40803ecc8abd3b0. I'll merge to master once the test builds at https://hydra.nixos.org/jobset/nixpkgs/haskell-updates have come back successfully.
